src/Pos/Wallet/Web/Server/Sockets.hs

```diff
@@ -10,64 +10,74 @@ module Pos.Wallet.Web.Server.Sockets
        , MonadWalletWebSockets
        , getWalletWebSockets
        , ConnectionsVar
-       , initWSConnection
-       , closeWSConnection
+       , initWSConnections
+       , closeWSConnections
        , upgradeApplicationWS
-       , notify
+       , notifyAll
        , runWalletWS
        ) where
 
 import           Universum
 
+import           Control.Concurrent.STM.TVar    (swapTVar)
 import           Data.Aeson                     (encode)
+import           Data.Default                   (Default (def))
+import qualified Data.IntMap.Strict             as IM
 import qualified Ether
 import           Network.Wai                    (Application)
 import           Network.Wai.Handler.WebSockets (websocketsOr)
 import qualified Network.WebSockets             as WS
 
 import           Pos.Aeson.ClientTypes          ()
 import           Pos.Wallet.Web.ClientTypes     (NotifyEvent (ConnectionClosed, ConnectionOpened))
+import           Pos.Wallet.Web.Error           (WalletError (..))
 
--- NODE: for now we are assuming only one client will be used. If there will be need for multiple clients we should extend and hold multiple connections here.
--- We might add multiple clients when we add user profiles but I am not sure if we are planning on supporting more at all.
-type ConnectionsVar = TVar (Maybe WS.Connection)
 
-initWSConnection :: MonadIO m => m ConnectionsVar
-initWSConnection = newTVarIO Nothing
+type WSConnection = WS.Connection
+type ConnectionsVar = TVar ConnectionMap
 
-closeWSConnection :: MonadIO m => ConnectionsVar -> m ()
-closeWSConnection var = liftIO $ do
-    conn <- readTVarIO var
-    atomically $ writeTVar var Nothing
-    maybe mempty (flip WS.sendClose ConnectionClosed) conn
+initWSConnections :: MonadIO m => m ConnectionsVar
+initWSConnections = newTVarIO def
 
-switchConnection :: ConnectionsVar -> WS.ServerApp
-switchConnection var pending = do
+closeWSConnection :: (MonadIO m, MonadThrow m) => ConnectionKey -> ConnectionsVar -> m ()
+closeWSConnection key var = liftIO $ do
+    maybeConn <- atomically $ deregisterConnection key var
+    case maybeConn of
+        Nothing   -> throwM $ InternalError "Attempted to close an unknown connection"
+        Just conn -> WS.sendClose conn ConnectionClosed
+
+closeWSConnections :: (MonadIO m, MonadThrow m) => ConnectionsVar -> m ()
+closeWSConnections var = liftIO $ do
+    conns <- atomically $ swapTVar var def
+    for_ (listConnections conns) $ flip WS.sendClose ConnectionClosed
+
+appendWSConnection :: ConnectionsVar -> WS.ServerApp
+appendWSConnection var pending = do
     conn <- WS.acceptRequest pending
-    closeWSConnection var
-    atomically . writeTVar var $ Just conn
-    sendWS var ConnectionOpened
+    key <- atomically $ registerConnection conn var
```

Comment:
Nu uslovno.
Following `sendWS` can then fail and `conn` will stuck in the map FOREVER